meshnets/data_processing/data_statistics.py

```diff
@@ -0,0 +1,79 @@
+"""File with functions to compute data statistics"""
+import numpy as np
+
+
+class OnlineStatistics:
+    """Computes mean ad standard deviation from streaming data
+
+    Args:
+      axis: Axis over which to compute. Default is None, which means
+        that the statistics are computed over the flattened array.
+
+    Example:
+
+    >>> dataset = [[1, 2], [3, 4]]
+    >>> stats = OnlineStatistics()
+    >>> for data in dataset:
+    ...     stats.update(data)
+    >>> stats.get_mean()
+    array([[2.5]])
+    >>> stats.get_std()
+    array([[1.11803399]])
+    >>> stats = OnlineStatistics(axis=0)
+    >>> for data in dataset:
+    ...     stats.update(data)
+    >>> stats.get_mean()
+    array([[2., 3.]])
+    >>> stats.get_std()
+    array([[1., 1.]])
+    >>> stats = OnlineStatistics(axis=1)
+    >>> for data in dataset:
+    ...     stats.update(data)
+    >>> stats.get_mean()
+    array([[1.5], [3.5]])
+    >>> stats.get_std()
+    array([[0.5], [0.5]])
+    """
+
+    def __init__(self, axis=None):
+        self.axis = axis
+
+        self.n = None
+        self.cumulative_sum = None
+
+        self.cumulative_sum_2 = None
+
+    def update(self, example):
+        data_point = np.expand_dims(example, 0)
+        self.sum_element = np.sum(data_point, axis=self.axis, keepdims=True)
+        self.sum_element_2 = np.sum(data_point**2,
+                                    axis=self.axis,
+                                    keepdims=True)
+        if self.cumulative_sum is None:
+            self.cumulative_sum = self.sum_element
+            self.cumulative_sum_2 = self.sum_element_2
+        else:
+            sum_element = np.sum(data_point, axis=self.axis, keepdims=True)
+            sum_element_2 = np.sum(data_point**2, axis=self.axis, keepdims=True)
+            self.cumulative_sum = np.sum(np.concatenate(
+                (self.cumulative_sum, sum_element)),
+                                         axis=self.axis,
+                                         keepdims=True)
+            self.cumulative_sum_2 = np.sum(np.concatenate(
+                (self.cumulative_sum_2, sum_element_2)),
+                                           axis=self.axis,
+                                           keepdims=True)
+        n = np.sum(np.ones_like(data_point), axis=self.axis, keepdims=True)
+        if self.n is None:
+            self.n = n
+        else:
+            self.n = np.sum(np.concatenate((self.n, n)),
```

Comment:
Should we be able to just add the self.n and n "pointwise"?

Reply:
Not sure what you mean by this

Reply:
What is the function actually trying to do?

Reply:
When computing the mean of array, in numpy, we can usually do something like:

```python
a = [[1, 2], [3, 4]]
np.mean(a)
# 2.5
np.mean(a, axis=0)
# [[2, 3]]
np.mean(a, axis=1)
# [[1.5], [3.5]]
```

However, our datasets do not fit into memory. So we have to perform this computation online. Thus we keep track of the total sum and the number of samples (`n`).
The extra code is just to keep the functionality of doing this along a given axis (just like in numpy).

We do the same for the standard deviation.

Reply:
Just thinking if it is not simple to add pointwise the `n` you computed and the previous one `self.n`. Like, with `np.add`?

Reply:
We could perhaps later improve on the `n` computation. At the moment we also keep track of it using the axis. That is, for the previous example:

```python
s = OnlineStatistics(axis=None)
s.update([1, 2])
s.n
# [2]
s.update([3, 4])
s.n
# [4]
###
s = OnlineStatistics(axis=0)
s.update([1, 2])
s.n
# [[1, 1]]
s.update([3, 4])
s.n
# [[2, 2]]
###
s = OnlineStatistics(axis=1)
s.update([1, 2])
s.n
# [[2]]
s.update([3, 4])
s.n
# [[2], [2]]
```